src/Domain/InterfaceHelpers.hpp

```diff
@@ -0,0 +1,119 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DirectionMap.hpp"
+#include "Domain/Direction.hpp"
+#include "Domain/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace InterfaceHelpers_detail {
+
+template <typename T, typename = cpp17::void_t<>>
+struct get_volume_tags_impl {
+  using type = tmpl::list<>;
+};
+template <typename T>
+struct get_volume_tags_impl<T, cpp17::void_t<typename T::volume_tags>> {
+  using type = typename T::volume_tags;
+};
+
+/// Retrieve `T::volume_tags`, defaulting to an empty list
+template <typename T>
+using get_volume_tags = tmpl::type_from<get_volume_tags_impl<T>>;
+
+template <typename Tag, typename DirectionsTag, typename VolumeTags>
+struct make_interface_tag_impl {
+  using type = tmpl::conditional_t<tmpl::list_contains_v<VolumeTags, Tag>, Tag,
+                                   ::Tags::Interface<DirectionsTag, Tag>>;
+};
+
+/// Add `::Tags::Interface` wrappers to the tags if they are not listed as being
+/// taken from the volume.
+template <typename Tag, typename DirectionsTag>
+using make_interface_argument_tags =
+    tmpl::transform<typename Tag::argument_tags,
+                    make_interface_tag_impl<tmpl::_1, tmpl::pin<DirectionsTag>,
+                                            tmpl::pin<get_volume_tags<Tag>>>>;
+
+/// Pull the direction's entry from interface arguments, passing volume
+/// arguments through unchanged.
+template <bool IsVolumeTag>
+struct unmap_interface_args;
+
+template <>
+struct unmap_interface_args<true> {
+  template <typename T>
+  using f = T;
+
+  template <size_t VolumeDim, typename T>
+  static constexpr const T& apply(const ::Direction<VolumeDim>& /*direction*/,
+                                  const T& arg) noexcept {
+    return arg;
+  }
+};
+
+template <>
+struct unmap_interface_args<false> {
+  template <typename T>
+  using f = typename T::mapped_type;
+
+  template <size_t VolumeDim, typename T>
+  static constexpr decltype(auto) apply(const ::Direction<VolumeDim>& direction,
+                                        const T& arg) noexcept {
+    return arg.at(direction);
+  }
+};
+
+template <typename DirectionsTag, typename VolumeTags,
+          typename InterfaceInvokable, typename DbTagsList,
+          typename... ArgumentTags, typename... ExtraArgs>
+SPECTRE_ALWAYS_INLINE constexpr auto interface_apply_impl(
+    InterfaceInvokable&& interface_invokable,
+    const db::DataBox<DbTagsList>& box, tmpl::list<ArgumentTags...> /*meta*/,
+    ExtraArgs&&... extra_args) noexcept {
+  using interface_return_type = std::decay_t<decltype(
+      interface_invokable(std::declval<db::item_type<ArgumentTags>>()...,
```

Comment:
I think this can be made to work with `db::BaseTag`s (at least for the volume tags).  See if adding the second argument to `db::item_type` does that.  If not don't worry about it.

Reply:
yep, works for volume tags